pennylane/numpy/wrapper.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module provides the PennyLane wrapper functions for modifying NumPy,
+such that it accepts the PennyLane :class:`~.tensor` class.
+"""
+from collections import Sequence
+import functools
+
+from autograd import numpy as _np
+
+from .tensor import tensor
+
+
+def extract_tensors(x):
+    """Iterate through an iterable, and extract any PennyLane
+    tensors that appear.
+    """
+    if isinstance(x, tensor):
+        # If the item is a tensor, return it
+        yield x
+    elif isinstance(x, Sequence) and not isinstance(x, (str, bytes)):
+        # If the item is a sequence, recursively look through its
+        # elements for tensors.
+        # NOTE: we choose to branch on Sequence here and not Iterable,
+        # as NumPy arrays are not Sequences.
+        for item in x:
+            yield from extract_tensors(item)
+
+
+def tensor_wrapper(obj):

```

Comment:
I found the name `obj` too obscure. Would we be better suited with a more descriptive name of what types we expect to be used (`func`?, `callable_obj`?)

Reply:
I've had a think about this, I don't think `func` or `callable_obj` are general enough, since this wrapper acts on both callables _and_ classes. The only objects it wouldn't explicitly work on are instances of non-callable classes. I'll add this to the docstring.

Reply:
```rest
Decorator that wraps callable objects and classes so that they both accept
a ``requires_grad`` keyword argument, as well as returning a PennyLane
:class:`~.tensor`.

Only if the decorated object returns an ``ndarray`` is the
output converted to a :class:`~.tensor`; this avoids superfluous conversion
of scalars and other native-Python types.

Args:
    obj: a callable object or class
```